File: internal/projection/nested_test.go
package projection

import "testing"

func nested(runID, nodeID string) *ParentRef {
	return &ParentRef{RunID: runID, NodeID: nodeID}
}

func nestedEvent(runID, graph string, step int, parent *ParentRef, frontier ...string) StepEvent {
	return StepEvent{
		RunID: runID, Graph: graph, Step: step,
		Frontier: frontier, At: at(step), Parent: parent,
	}
}

// A nested run keeps its own level sequence. Folding it into its parent's would corrupt the
// counter this projection uses to reject stale events.
func TestANestedRunIsARunOfItsOwn(t *testing.T) {
	p := New()
	_, _, _ = p.Apply(nestedEvent("parent", "top", 1, nil, "sub"))

	if _, _, err := p.Apply(nestedEvent("child", "inner", 1, nested("parent", "sub"), "c2")); err != nil {
		t.Fatalf("Apply: %v", err)
	}

	if got := len(p.List()); got != 2 {
		t.Fatalf("the projection holds %d runs, want 2", got)
	}
	child, ok := p.Get("child")
	if !ok {
		t.Fatal("the nested run is unknown")
	}
	if child.Parent == nil || child.Parent.NodeID != "sub" {
		t.Errorf("Parent = %+v, want the node it belongs to", child.Parent)
	}
	if parent, _ := p.Get("parent"); parent.Step != 1 {
		t.Errorf("the parent's step = %d, want 1 — a child must not advance it", parent.Step)
	}
}

// The only question the interface has: what ran inside this node?
func TestChildOfFindsTheRunBelongingToANode(t *testing.T) {
	p := New()
	_, _, _ = p.Apply(nestedEvent("child", "inner", 1, nested("parent", "sub"), "c2"))
	_, _, _ = p.Apply(nestedEvent("other", "inner", 1, nested("parent", "autre"), "x"))

	got, ok := p.ChildOf("parent", "sub")
	if !ok {
		t.Fatal("no nested run found for the node")
	}
	if got.ID != "child" {
		t.Errorf("ChildOf = %q, want child", got.ID)
	}

	if _, ok := p.ChildOf("parent", "jamais"); ok {
		t.Error("a node that ran no subgraph reported one")
	}
}

// The same node running twice — a retry, a loop — is two nested runs. The freshest is the
// one worth drawing.
func TestChildOfPrefersTheMostRecentRun(t *testing.T) {
	p := New()
	_, _, _ = p.Apply(nestedEvent("first", "inner", 1, nested("parent", "sub"), "a"))
	_, _, _ = p.Apply(nestedEvent("second", "inner", 5, nested("parent", "sub"), "b"))

	got, _ := p.ChildOf("parent", "sub")
	if got.ID != "second" {
		t.Errorf("ChildOf = %q, want the later run", got.ID)
	}
}

func TestApplyRejectsAParentThatPointsNowhere(t *testing.T) {
	cases := map[string]*ParentRef{
		"no run id":  {NodeID: "sub"},
		"no node id": {RunID: "parent"},
		"blank":      {RunID: "  ", NodeID: "sub"},
	}
	for name, parent := range cases {
		t.Run(name, func(t *testing.T) {
			if _, _, err := New().Apply(nestedEvent("child", "inner", 1, parent, "a")); err == nil {
				t.Error("accepted a parent reference that points nowhere")
			}
		})
	}
}

File: web/src/runs/nested.test.ts
import { childRunOf, topLevelRuns } from './nested'
import type { Run } from './types'

function run(id: string, over: Partial<Run> = {}): Run {
  return {
    id,
    graph: 'g',
    status: 'running',
    step: 1,
    frontier: [],
    started_at: '2026-07-28T12:00:00Z',
    updated_at: '2026-07-28T12:00:00Z',
    ...over,
  }
}

describe('the runs worth listing', () => {
  // A nested run is already drawn inside the node that produced it. Listing it again would
  // show the same work twice and hide which of the two is the whole story.
  it('leaves nested runs out of the list', () => {
    const runs = [run('a'), run('b', { parent: { run_id: 'a', node_id: 'sub' } })]

    expect(topLevelRuns(runs).map((r) => r.id)).toEqual(['a'])
  })

  it('keeps every run when none is nested', () => {
    expect(topLevelRuns([run('a'), run('b')])).toHaveLength(2)
  })
})

describe('the run inside a node', () => {
  it('finds the run a subgraph node produced', () => {
    const runs = [run('a'), run('child', { parent: { run_id: 'a', node_id: 'sub' } })]

    expect(childRunOf(runs, 'a', 'sub')?.id).toBe('child')
  })

  it('finds nothing for a node that ran no subgraph', () => {
    const runs = [run('a'), run('child', { parent: { run_id: 'a', node_id: 'sub' } })]

    expect(childRunOf(runs, 'a', 'autre')).toBeUndefined()
  })

  it('does not confuse two parents', () => {
    const runs = [
      run('x', { parent: { run_id: 'a', node_id: 'sub' } }),
      run('y', { parent: { run_id: 'b', node_id: 'sub' } }),
    ]

    expect(childRunOf(runs, 'b', 'sub')?.id).toBe('y')
  })

  // A node running its subgraph twice is two runs; the freshest is what is happening now.
  it('prefers the most recent execution', () => {
    const runs = [
      run('first', {
        parent: { run_id: 'a', node_id: 'sub' },
        started_at: '2026-07-28T12:00:00Z',
      }),
      run('second', {
        parent: { run_id: 'a', node_id: 'sub' },
        started_at: '2026-07-28T13:00:00Z',
      }),
    ]

    expect(childRunOf(runs, 'a', 'sub')?.id).toBe('second')
  })
})
